File: servicenow-samples/06-inbound-python/snow_to_cf_inbound/app/callbacks/change_tracker.py
"""In-memory change tracker, same role as the CW / Halo inbound samples.

Journal entries (`sys_journal_field`) carry no incrementing integer id, so we
track the latest `sys_created_on` we've forwarded per incident. That column is
stored UTC in the sortable `yyyy-MM-dd HH:mm:ss` format, so a plain string
compare is a correct "newer than" test.
"""

from __future__ import annotations

from dataclasses import dataclass
from threading import Lock


@dataclass(frozen=True)
class IncidentState:
    last_journal_created_on: str
    closed: bool


class ChangeTracker:
    def __init__(self) -> None:
        self._state: dict[str, IncidentState] = {}
        self._lock = Lock()

    def snapshot(self, sys_id: str) -> IncidentState:
        with self._lock:
            return self._state.get(
                sys_id, IncidentState(last_journal_created_on="", closed=False)
            )

    def record(self, sys_id: str, state: IncidentState) -> None:
        with self._lock:
            self._state[sys_id] = state

File: servicenow-samples/04-inbound-webapp/ContraForce.Samples.SnowInbound/CallbackModels/ChangeTracker.cs
using System.Collections.Concurrent;

namespace ContraForce.Samples.SnowInbound.CallbackModels;

/// <summary>
/// Tracks the last processed state per ServiceNow incident (keyed by sys_id)
/// so that repeated Business Rule callbacks — ServiceNow fires one on every
/// update — don't generate duplicate comments on the ContraForce side.
/// </summary>
/// <remarks>
/// Journal entries (<c>sys_journal_field</c>) have no incrementing integer id,
/// so we track the latest <c>sys_created_on</c> we've forwarded. That column is
/// stored UTC in the sortable <c>yyyy-MM-dd HH:mm:ss</c> format, so a plain
/// ordinal string compare is a correct "newer than" test.
///
/// This sample keeps state in memory. For multi-instance deployments, replace
/// with a Redis, Cosmos, or SQL-backed implementation.
/// </remarks>
public sealed class ChangeTracker
{
    private readonly ConcurrentDictionary<string, IncidentState> _state = new(
        StringComparer.Ordinal
    );

    public IncidentState Snapshot(string sysId) =>
        _state.TryGetValue(sysId, out var s)
            ? s
            : new IncidentState(LastJournalCreatedOn: string.Empty, Closed: false);

    public void Record(string sysId, IncidentState state) => _state[sysId] = state;
}

public sealed record IncidentState(string LastJournalCreatedOn, bool Closed);
